rfcs/20200422-concurrent-global-scratchpad-mode/README.md

```diff
@@ -0,0 +1,185 @@
+2. `global_scratchpad_t` --> `thread_private_scratchpad_t`, thereby emphasizing
+   that such scratchpad storage is tied to a particular thread of execution, and
+   can be shared by any primitives executed by this thread, which enables
+   reduced memory footprint.
+
+
+These types, as well as the subsequently proposed additions to the library, are
+expected to be more self explanatory for the user.
+
+## API
+
+The scope of this proposal includes introducing two new `scratchpad_mode`
+primitive attributes:
+
+1. `primitive_private`. In this case, the user enforces all primitives to use an
+   independent scratchpad. The user may use this flag for thread safety, and it
+   enables the user to estimate the total scratchpad memory consumption,
+   although can increase memory consumption dramatically;
+
+2. `thread_private`. In this case, the user enforces all primitives to use a
+   shared global scratchpad. Thread safety is not guaranteed here, however the
+   user may work around this if necessary. This flag can significantly reduce
+   the memory consumption due to scratchpad, and the user can query the total
+   memory usage due to scratchpad from the last primitive.
+
+These flags will work on a primitive by primitive basis, meaning that the
+primitives will use different storage if different flags were used for their
+instantiation. This enables a flexible decision making process for the user with
+respect to prioritizing thread safety and total memory allocation, instead of
+being enforced to the current scratchpad implementation in oneDNN.
+
+It is important to note that, by using the above flags, the user enforces the
+corresponding behavior discussed above, rather than serving as a hint.
+
+The following values for `dnnl_scratchpad_mode_t` will be introduced:
+
+```cpp
+/// The user enforces the primitive-private scratchpad mode for the primitive.
+dnnl_scratchpad_primitive_private,
+/// The user enforces the thread-private scratchpad mode for the primitive.
+dnnl_scratchpad_thread_private,
+```
+
+Similarly, the following `scratchpad_mode` enum values will be introduced:
+
+```cpp
+/// Scratchpad mode
+///
+primitive_private = dnnl_scratchpad_primitive_private,
+///
+thread_private = dnnl_scratchpad_thread_private,
+```
+
+### Usage models
+
+Since the new scratchpad modes are intended to complement existing one instead
+of replacing them, new usage models are assumed to be favorable for the user in
+a particular instance. These usage models are briefly presented in the table
+below:
+
+| Scratchpad mode       | Use case | Advantages |
+|-----------------------|----------|------------|
+| **library**           | The library allocates scratchpad storage.<br>The mode is chosen by primitive implementation.                    | No API usage<br>     |
+| **primitive_private** | The library allocates scratchpad storage.<br>New scratchpad storage is created for each primitive.              | Thread-safety        |
```

Comment:
I find mentioning thread-safety here confusing because if a primitive was created for this scratchpad mode then it cannot be executed by multiple threads simultaneously. You probably meant that the primitive can be created by one thread, but executed by another. Is that what's implied here?

Reply:
Indeed that's what I was implying, I will make this more explicit throughout the document.